gsw_oceanographic_toolbox.c

```diff
@@ -3933,6 +3933,352 @@ p_sequence(double p1, double p2, double max_dp_i, double *pseq, int *nps)
+!
+!  Calculates dynamic height anomaly as the integral of specific volume
+!  anomaly from the pressure p of the bottle to the reference pressure
+!  p_ref.
+!
+!  Hence, geo_strf_dyn_height is the dynamic height anomaly with respect
+!  to a given reference pressure.  This is the geostrophic streamfunction
+!  for the difference between the horizontal velocity at the pressure
+!  concerned, p, and the horizontal velocity at p_ref.  Dynamic height
+!  anomaly is the geostrophic streamfunction in an isobaric surface.  The
+!  reference values used for the specific volume anomaly are
+!  SSO = 35.16504 g/kg and CT = 0 deg C.  This function calculates
+!  specific volume anomaly using the computationally efficient
+!  expression for specific volume of Roquet et al. (2015).
+!
+!  This function evaluates the pressure integral of specific volume using
+!  SA and CT linearly interpolated with respect to pressure.  Additional
+!  interpolation methods may be added or substituted.
+!
+!  SA    =  Absolute Salinity                                      [ g/kg ]
+!  CT    =  Conservative Temperature (ITS-90)                     [ deg C ]
+!  p     =  sea pressure                                           [ dbar ]
+!           ( i.e. absolute pressure - 10.1325 dbar )
+!  p_ref =  reference pressure                                     [ dbar ]
+!           ( i.e. reference absolute pressure - 10.1325 dbar )
+!
+!  geo_strf_dyn_height  =  dynamic height anomaly               [ m^2/s^2 ]
+!   Note. If p_ref falls outside the range of a
+!     vertical profile, the dynamic height anomaly for each bottle
+!     on the whole vertical profile is returned as NaN.
+!--------------------------------------------------------------------------
+*/
+
+/*
+    Make a new grid based on an original monotonic array, typically P, such that
+    on the new monotonic grid, p_i:
+
+        1) The first element is p[0], the last is p[nz-1].
+        2) Approximate integer multiples of dp are included.
+        3) All original p points are included.
+        4) The value p_ref is included.
+
+    This function allocates memory; it is
+    the responsibility of subsequent code to free this memory.
+
+    In addition to the new p_i grid, the function returns the array
+    of indices (p_indices)
+    of the original p grid points in p_i, and a pointer to
+    the index of p_ref within p_i.
+
+    Its return argument is the number of elements in p_i.
+*/
+static int refine_grid_for_dh(double *p, double p_ref, int nz,
+    double dp,
+    double *p_i, int ni_max,  /* size of p_i array; larger than needed */
+    int *p_indices, int *p_ref_ind_ptr)
+{
+    int i, iuniform, iorig;
+    double p_next;
+    /* Don't add a new point if it is within p_tol of an original. */
+    double p_tol = 0.001 * dp;
+
+    p_i[0] = p[0];
+    p_indices[0] = 0;
+    *p_ref_ind_ptr = -1;  /* initialize to a flag value */
+    if (p_ref <= p[0] + p_tol)
+    {
```

Comment:
Eric, you asked about thoughts on the indentation style.  I think any style is okay, so long as it's consistent within a function. Any reasonable C programmer will be familiar with a variety of styles.  (I would have used KR, but the style you've used is perfectly fine.) I am viewing this in a browser, and I can't see if you are using TAB characters for spacing. I've never been keen on that because it makes the code look different with different editor settings ... I realize that's part of the appeal, though, and if you've done that, I wouldn't suggest a reformat.